File: LinkedLists/Circular-Singly-LinkedList/CircSinglyLinkedList.py
class Node:

    def __init__(self):
        self.data = None
        self.Next = None

class CircSinglyLinkedList:

    def __init__(self):
        self.head = None
        self.size = 0

    #Check whether the list is Empty or not
    def isEmpty(self):
        return self.head == None

    #Traversing The List
    def traverse(self):

        if self.isEmpty():
            print("List is Empty!");
            return

        currentNode = self.head

        for i in range(self.size):
            print(currentNode.data, end = " -> ")
            currentNode = currentNode.Next

        print()


    #Inserting Node(s)
    def insert(self, x, loc):

        newNode = Node()
        newNode.data = x

        if loc == 0:

            if self.isEmpty():

                self.head = newNode
                newNode.Next = self.head

            else:

                newNode.Next = self.head
                self.head = newNode

                currentNode = self.head

                for i in range(self.size):
                    currentNode = currentNode.Next

                currentNode.Next = self.head

        else:

            if loc > self.size:

                print("The list is only ", size, " nodes")
                return

            currentNode = self.head

            for i in range(loc - 1):
                currentNode = currentNode.Next

            newNode.Next = currentNode.Next
            currentNode.Next = newNode

        self.size += 1


    #Deleting Node(s)
    def delete(self, loc):

        if self.isEmpty():

            print("List is Empty!")
            return

        if loc > self.size - 1:

            print("The list is only ", self.size, "nodes")
            return

        currentNode = self.head
        previousNode = self.head

        if loc == 0:

            if self.size == 1:

                self.head = None

            else:

                self.head = currentNode.Next

                for i in range(self.size - 1):
                    currentNode = currentNode.Next

                currentNode.Next = self.head

        elif loc == self.size - 1:

            for i in range(loc - 1):
                currentNode = currentNode.Next

            currentNode.Next = self.head

        else:

            for i in range(loc):

                currentNode = currentNode.Next

                if i == 0:
                    continue

                previousNode = previousNode.Next
            
            previousNode.Next = currentNode.Next

        self.size -= 1


c = CircSinglyLinkedList()

c.traverse();

c.insert(1, 0);
c.traverse();

c.insert(2, 1);
c.traverse();

c.insert(3, 2);
c.traverse();

c.insert(5, 3);
c.traverse();

c.insert(4, 3);
c.traverse();

#c.insert(-1, 0);
#c.traverse();

c.delete(0);
c.traverse();

c.delete(3);
c.traverse();

c.delete(1);
c.traverse();

c.delete(0);
c.traverse();

c.delete(1);
c.traverse();

c.delete(0);
c.traverse();

File: LinkedLists/Circular-Singly-LinkedList/CircSinglyLinkedList.java
class CircSinglyLinkedList{

    //Establishing Node
    class Node {
        int data;
        Node next;
    }

    private Node head;
    private int size;

    //Initialization
    CircSinglyLinkedList(){

        head = null;
        size = 0;

    }

    //Check whether the list is Empty or not
    public boolean isEmpty(){
        return head == null;
    }

    //Traversing The List
    public void traverse(){

        if(isEmpty()){
            System.out.println("List is Empty!\n");
            return;
        }

        Node currentNode = head;

        do {

            System.out.print(currentNode.data + " -> ");
            currentNode = currentNode.next;

        } while(currentNode != head);

        System.out.println();

    }

    //Inserting Node(s)
    void insert(int x, int loc){

        Node newNode = new Node();
        newNode.data = x;

        if(loc == 0){

            if(isEmpty()){

                head = newNode;
                newNode.next = head;

            } else {

                newNode.next = head;
                head = newNode;

                Node currentNode = head;

                for(int i = 0; i < size; i++)
                    currentNode = currentNode.next;

                currentNode.next = head;

            }

        } else {

            if(loc > size){

                System.out.println("The list is only " + size + " nodes!");
                return;

            }

            Node currentNode = head;

            for(int i = 0; i < loc - 1; i++)
                currentNode = currentNode.next;

            newNode.next = currentNode.next;
            currentNode.next = newNode;

        }

        size++;

    }

    //Deleting Node(s)
    void delete(int loc){

        if(isEmpty()){

            System.out.println("List is Empty!\n");
            return;

        }

        if(loc > size - 1){

            System.out.println("The list is only " + size + " nodes!");
            return;

        }

        Node currentNode = head;
        Node previousNode = head;

        if(loc == 0){

            if(size == 1)
                head = null;
            else {

                head = currentNode.next;

                for(int i = 0; i < size - 1; i++)
                    currentNode = currentNode.next;

                currentNode.next = head;

            }

        } else if(loc == size - 1){

            for(int i = 0; i < loc - 1; i++)
                currentNode = currentNode.next;

            currentNode.next = head;

        } else {

            for(int i = 0; i < loc; i++){

                currentNode = currentNode.next;

                if(i == 0)
                    continue;

                previousNode = previousNode.next;

            }

            previousNode.next = currentNode.next;

        }

        size--;

    }

}
